drivers/i2c/i2c_wch.c

```diff
@@ -0,0 +1,498 @@
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	atomic_set(&data->xfer_done, 0);
+
+	data->current.msg = msg;
+	data->current.idx = 0U;
+	data->current.err = 0U;
+	data->current.addr = addr;
+
+	regs->CTLR1 |= I2C_CTLR1_PE;
+	regs->CTLR1 |= I2C_CTLR1_ACK;
+
+	if (msg->flags & I2C_MSG_RESTART) {
+		if (regs->CTLR1 & I2C_CTLR1_STOP) {
+			regs->CTLR1 &= ~I2C_CTLR1_STOP;
+		}
+		regs->CTLR1 |= I2C_CTLR1_START;
+	}
+}
+
+static int32_t wch_i2c_msg_end(const struct device *dev)
+{
+	struct i2c_wch_data *data = dev->data;
+
+	if (!data->current.err) {
+		return 0;
+	}
+
+	if (data->current.err & (I2C_STAR1_ARLO >> 8)) {
+		LOG_DBG("ARLO");
+	}
+
+	if (data->current.err & (I2C_STAR1_AF >> 8)) {
+		LOG_DBG("NACK");
+	}
+
+	if (data->current.err & (I2C_STAR1_BERR >> 8)) {
+		LOG_DBG("ERR");
+	}
+
+	data->current.err = 0;
+
+	return -EIO;
+}
+
+static void wch_i2c_config_interrupts(I2C_TypeDef *regs, bool enable)
+{
+	if (enable) {
+		regs->CTLR2 |= (I2C_CTLR2_ITERREN | I2C_CTLR2_ITEVTEN | I2C_CTLR2_ITBUFEN);
+	} else {
+		regs->CTLR2 &= ~(I2C_CTLR2_ITERREN | I2C_CTLR2_ITEVTEN | I2C_CTLR2_ITBUFEN);
+	}
+}
+
+static int32_t wch_i2c_begin_transfer(const struct device *dev, struct i2c_msg *msg, uint16_t addr)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	wch_i2c_msg_init(dev, msg, addr);
+
+	wch_i2c_config_interrupts(regs, true);
+
+	k_sched_lock();
+	while (atomic_get(&data->xfer_done) == 0) {
+	}
+	k_sched_unlock();
+
+	return wch_i2c_msg_end(dev);
+}
+
+static void wch_i2c_finish_transfer(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	I2C_TypeDef *regs = config->regs;
+
+	wch_i2c_config_interrupts(regs, false);
+
+	regs->CTLR1 &= ~I2C_CTLR1_PE;
+}
+
+static void wch_i2c_configure_timing(I2C_TypeDef *regs, uint32_t clock_rate, uint32_t bitrate)
+{
+	uint16_t clock_config;
+	uint16_t freq_range;
+
+	if (bitrate <= I2C_BITRATE_STANDARD) {
+		clock_config = MAX((uint16_t)(clock_rate / (bitrate << 1)), 4);
+	} else {
+		clock_config = MAX((uint16_t)(clock_rate / (bitrate * 3)), 1) | I2C_CKCFGR_FS;
+	}
+
+	regs->CKCFGR = clock_config;
+
+	freq_range = regs->CTLR2;
+	freq_range &= ~I2C_CTLR2_FREQ;
+	freq_range |= (uint16_t)(clock_rate / 1000000);
+
+	regs->CTLR2 = freq_range;
+}
+
+static int i2c_wch_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_wch_config *config = dev->config;
+	I2C_TypeDef *regs = config->regs;
+	clock_control_subsys_t clk_sys;
+	uint32_t clock_rate;
+	int err;
+
+	clk_sys = (clock_control_subsys_t)(uintptr_t)config->clk_id;
+
+	err = clock_control_get_rate(config->clk_dev, clk_sys, &clock_rate);
+	if (err != 0) {
+		return err;
+	}
+
+	if (!(I2C_MODE_CONTROLLER & dev_config)) {
+		return -EINVAL;
+	}
+
+	regs->CTLR1 &= ~I2C_CTLR1_PE;
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
```

Comment:
wch_i2c_configure_timing supports more modes than listed here. Call wch_i2c_configure_timing directly and let it return an error on unsupported.

I'd also drop the I2C_SPEED_STANDARD -> 100'000 conversion, as wch_i2c_configure_timing then interprets the parameter as a I2C_SPEED_STANDARD etc, so there's more of a chance of introducing bugs through a bad conversion.

Reply:
Done